### PR TITLE
ClientV2 Links

### PR DESCRIPTION
**Pull request by mfedderly**
Adds support for traversing links from objects.

```typescript
// for links without multiplicity we support the lookup directly
employee.$link.lead.get();

// for links with multiplicity you can either
lead.$link.peeps.fetchPageOrThrow() // fetch paged results
lead.$link.peeps.get(peepId) // or get a linked object by its PK directly
```

All three calls support being passed an options argument with a list of property names to get only a subset of the original object's properties.

### CODE REVIEW

File: packages/client/src/definitions/LinkDefinitions.ts [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+export type OsdkObjectLinksEntry<
+  K extends ObjectTypeKeysFrom<O>,
+  O extends OntologyDefinition<any>,
+  L extends ObjectTypeLinkKeysFrom<O, K>,
+> = ObjectTypeLinkDefinitionFrom<O, K, L> extends { multiplicity: false } ? {
+    /** Load the linked object */
+    get: () => OsdkObjectFrom<
+      ObjectTypeLinkDefinitionFrom<O, K, L>["targetType"],
+      O
+    >;
```
Comment (mfedderly) [inline packages/client/src/definitions/LinkDefinitions.ts:49]: needs to support `select` option

File: packages/client/src/object/convertWireToOsdkObjects.ts [hunk 2]
```diff
@@ -64,6 +122,13 @@ export function convertWireToOsdkObjects<
   for (const obj of objs) {
     Object.setPrototypeOf(obj, proto);
 
+    Object.defineProperty(obj, OriginClient, {
```
Comment (mfedderly) [inline packages/client/src/object/convertWireToOsdkObjects.ts:125]: I considered pushing this into the proto but then it made the caching semantics of `createCachedOntologyTransform` worse. Right now it caches on `===` ObjectTypeDefinition, which actually caches pretty well across client recreations and such.

File: packages/client/src/object/fetchPageOrThrow.ts [hunk 3]
```diff
@@ -69,13 +70,17 @@ export async function fetchPageOrThrow<
...
+    select: ((args?.select as string[] | undefined) ?? []), // FIXME?
   };
 
   if (args?.nextPageToken) {
     body.pageToken = args.nextPageToken;
   }
 
+  if (args?.pageSize != null) {
+    body.pageSize = args.pageSize;
+  }
```
Comment (mfedderly) [inline packages/client/src/object/fetchPageOrThrow.ts:82]: Was there a reason this wasn't exposed before?

File: packages/client/src/object/getLinkedObjectByPkOrThrow.ts [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+    linkTypeApiName,
+    linkedObjectPrimaryKey,
+    {
+      select: [],
+    },
+  );
+
+  return convertWireToOsdkObjects(
+    client,
+    client.ontology.objects[sourceApiName].links[linkTypeApiName].targetType,
```
Comment (mfedderly) [inline packages/client/src/object/getLinkedObjectByPkOrThrow.ts:47]: Do we actually want to be passing in apiName here? Or just grab `__apiName` from each object we convert?

File: packages/client/src/object/getLinkedObjectOrThrow.ts [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+    { pageSize: 1 },
+  );
+
+  if (result.data.length !== 1 || result.nextPageToken != null) {
+    throw new PalantirApiError(
+      `Expected a single result but got ${result.data.length} instead${
+        result.nextPageToken != null ? " with nextPageToken set" : ""
+      }`,
+    );
+  }
```
Comment (mfedderly) [inline packages/client/src/object/getLinkedObjectOrThrow.ts:44]: ^ is this how we want to handle non-conjure errors related to network requests?

File: packages/client/src/object/convertWireToOsdkObjects.ts [hunk 6]
```diff
@@ -36,6 +43,57 @@ function createPrototype<
...
+                }),
+            };
+          }
+        },
+      });
+    },
+    enumerable: false,
+    configurable: false,
+    writable: false,
+  });
```
Comment (mfedderly) [inline packages/client/src/object/convertWireToOsdkObjects.ts:95]: ^ these methods need runtime code testing before I feel comfortable shipping it

File: packages/client/src/object/convertWireToOsdkObjects.ts [hunk 7]
```diff
@@ -36,6 +43,57 @@ function createPrototype<
   const objDef = ontology.objects[type];
   const proto = {};
 
+  Object.defineProperty(proto, "$links", {
+    get: function() {
+      const client = this[OriginClient] as ClientContext<O>;
```
Comment (ericanderson) [inline packages/client/src/object/convertWireToOsdkObjects.ts:48]: Im happy to see this.

File: packages/client/src/object/fetchPageOrThrow.ts (hunk 3 above)
Comment (ericanderson) [reply packages/client/src/object/fetchPageOrThrow.ts:82]: Mistake on my part. Not intentional

File: packages/client/src/object/getLinkedObjectByPkOrThrow.ts (hunk 4 above)
Comment (ericanderson) [reply packages/client/src/object/getLinkedObjectByPkOrThrow.ts:47]: I am guessing that its equiv for now. In fact you could add an invariant check here until interface support that says that we expect them to be the same

File: packages/client/src/object/getLinkedObjectOrThrow.ts (hunk 5 above)
Comment (ericanderson) [reply packages/client/src/object/getLinkedObjectOrThrow.ts:44]: I think it doesn't matter in this case. We aren't using any of the other fields. We aren't leveraging retry (nor would i expect doing it again to work). 

Using this vs Error doesnt change anything besides how someone consumes it, right?

File: packages/client/src/object/getLinkedObjectByPkOrThrow.ts (hunk 4 above)
Comment (mfedderly) [reply packages/client/src/object/getLinkedObjectByPkOrThrow.ts:47]: (OR you stop having to look up the API name entirely, which also matches the legacy-client behavior)

**Review by ericanderson — APPROVED** [review]

File: packages/client/src/definitions/LinkDefinitions.test.ts [hunk 8]
```diff
@@ -0,0 +1,107 @@
...
+            Todos: {
+              get: <
+                A extends SelectArg<
+                  typeof MockOntology,
+                  "Todo",
+                  ObjectOrInterfacePropertyKeysFrom<
+                    typeof MockOntology,
+                    "Todo"
+                  >
+                >,
```
[the remark this answers is not in the document]
Comment (mfedderly) [reply packages/client/src/definitions/LinkDefinitions.test.ts:48]: Yeah I had to build this whole test out piecemeal, if it breaks hopefully its easy to keep diffing it forward.

**Pull request merged**